hs_labels/models.py

```diff
@@ -199,113 +233,157 @@ def clean_label(name):
 
     def label_resource(self, this_resource, this_label):
         """
-        assign a label to a resource
-
-        :param this_resource: Resource to be assigned a label.
-        :param this_label: Label to assign to resource
+        Assign a label to a resource
 
         Users are allowed to label any resource, including resources to which they do not have access.
         This is not an access control problem because labeling information is private.
         """
+        if __debug__:  # during testing only, check argument types and preconditions
+            assert isinstance(this_resource, BaseResource)
+            assert isinstance(this_label, basestring)
 
         # remove leading and trailing spaces
         label_string = UserLabels.clean_label(this_label)
-        UserResourceLabels.objects.get_or_create(resource=this_resource,
-                                                 label=label_string,
-                                                 user=self.user)
+        with transaction.atomic():  # empirically, get_or_create is not atomic.
```

Comment:
Is there unit test to show that using ```transaction.atomic()``` serves the purpose? I am not sure what benefits we get using ```transaction.atomic()``` here.